base/sort.jl

```diff
@@ -1323,15 +1386,34 @@ defalg(v::AbstractArray{Union{}}) = DEFAULT_UNSTABLE # for method disambiguation
 """
     sort!(v; alg::Algorithm=defalg(v), lt=isless, by=identity, rev::Bool=false, order::Ordering=Forward)
 
-Sort the vector `v` in place. A stable algorithm is used by default. You can select a
-specific algorithm to use via the `alg` keyword (see [Sorting Algorithms](@ref) for
-available algorithms). The `by` keyword lets you provide a function that will be applied to
-each element before comparison; the `lt` keyword allows providing a custom "less than"
-function (note that for every `x` and `y`, only one of `lt(x,y)` and `lt(y,x)` can return
-`true`); use `rev=true` to reverse the sorting order. These options are independent and can
-be used together in all possible combinations: if both `by` and `lt` are specified, the `lt`
-function is applied to the result of the `by` function; `rev=true` reverses whatever
-ordering specified via the `by` and `lt` keywords.
+Sort the vector `v` in place. A stable algorithm is used by default. A specific
+algorithm can be selected via the `alg` keyword (see [Sorting Algorithms](@ref)
+for available algorithms). Elements are first transformed by the function `by`
+and then compared according to either the function `lt` or the ordering
+`order`. Finally, the resulting order is reversed if `rev=true`.
+
```

Comment:
Maybe add something along the lines of:

'The transformation using the function `by` is done before each comparison.'

I think this behavior should be documented, as it has enormous performance implications for expensive `by`- functions, and differs from [sort in Python](https://docs.python.org/3/library/stdtypes.html#list.sort), where it is guaranteed that the `key`-function is called exactly once for each element.

Reply:
I've updated the `sort!` docstring to include the following:

>  The current implemention applies the `by` transformation before each comparison rather than once per element.